compiler/rustc_hir_typeck/src/errors.rs

```diff
@@ -321,15 +321,15 @@ impl HelpUseLatestEdition {
 }
 
 #[derive(Diagnostic)]
-#[diag(hir_typeck_invalid_callee, code = "E0618")]
+#[diag(hir_typeck_invalid_callee, code = E!(0618))]
 pub struct InvalidCallee {
     #[primary_span]
     pub span: Span,
     pub ty: String,
 }
 
 #[derive(Diagnostic)]
-#[diag(hir_typeck_int_to_fat, code = "E0606")]
+#[diag(hir_typeck_int_to_fat, code = E!(0606))]
```

Comment:
I think the strings are nicer to read here and would prefer if the macro kept taking strings and parsed them internally

Reply:
the proc macro could also take an integer if it's already parsing them anyways :grin:

Reply:
Tidy searches for occurrences of error codes, so it's simpler if all the error codes have a consistent form.

Reply:
Come to think of it, a function `E(0123)` should be doable instead of a macro `E!(0123)`, which is visually nicer, though it might require a few more imports of `E` via `use` items.

Either way, I have a strong opinion that a consistent, non-string-literal syntactic form should be used for all error codes appearing in the code. The `E!(0123)` and `E(0123)` forms are the only alternatives found so far that satisfy that.

Reply:
> Come to think of it, a function `E(0123)` should be doable instead of a macro `E!(0123)`, which is visually nicer, though it might require a few more imports of `E` via `use` items.

At that point we could just make all the error codes be ZSTs and use them directly as `E0123`

Reply:
Looking at this from the perspective of someone who doesn't work on rustc, it is not rare to grep for error codes in rustc to look at what is happening once in a while. It is most likely how many people who contribute to rustc with diagnostic improvements start. And those are just a fraction of people who would actually do the searching.

On one hand I agree with the sentiment of storing integers as Strings is silly. 

On the other rustc is already (in)famous for being hard to find things in. 

Not saying that this shouldn't happen, but this is probably a more impacting change than possibly expected. Though my guess is that once merged, I would be surprised if any confusion is voiced here. Just people at their desk surprised to see a change. Or someone who's never looked at the repo before confused on how to find an error code. So I just feel like I need to speak for those who do not look at this repo.

Reply:
I agree that it would be nice to keep the literal `E2234` in the source. I don't really care what's around it.

Reply:
> At that point we could just make all the error codes be ZSTs and use them directly as E0123

Make each code a separate type? That won't work, there are several hash tables storing error codes, so there must be a single type.

But making a constant `const Ennnn: ErrCode` for every code should be doable. And we could do something like `use rustc_error_codes::*` to avoid having to import multiple constants into the larger files. I can rework this PR to do that if people agree it's worthwhile.